plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +302,243 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+			g_autoptr(GsPluginEvent) event = gs_plugin_event_new ();
+			gs_utils_error_convert_gdbus (&local_error);
+			gs_plugin_event_set_app (event, app);
+			gs_plugin_event_set_error (event, local_error);
+			gs_plugin_event_add_flag (event, GS_PLUGIN_EVENT_FLAG_WARNING);
+			gs_plugin_report_event (plugin, event);
+		}
+
+		break;
+	}
+	case EOS_UPDATER_STATE_FETCHING: {
+		guint64 total_size = 0;
+		guint64 downloaded = 0;
+		gfloat progress = 0;
+
+		/* ensure the state transition to 'installing' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+		downloaded = eos_updater_get_downloaded_bytes (priv->updater_proxy);
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
+
+		if (total_size == 0) {
+			g_debug ("OS upgrade %s total size is 0!",
+				 gs_app_get_unique_id (app));
+		} else {
+			/* set progress only up to 90%, leaving the remaining
+			 * for applying the update */
+			progress = (gfloat) downloaded / (gfloat) total_size * 90.0;
+		}
+		gs_app_set_progress (app, (guint) progress);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_READY: {
+		/* if there's an update ready to be deployed and we are not yet
+		 * showing the OS upgrade as downloading (state 'installing'), then
+		 * just show it as available so the user has a change to click 'download'
+		 * which will end up deploying the update */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING) {
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+			break;
+		}
+
+		gs_app_set_progress (app, 95);
+
+		/* if the OS upgrade is being installed, but the updater state is
+		 * 'update-ready', it means we should proceed to applying the upgrade */
+		eos_updater_call_apply (priv->updater_proxy, NULL, NULL, NULL);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_APPLYING_UPDATE: {
+		/* ensure the state transition to 'installing' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+
+		/* set as 'installing' because if it is applying the update, we
+		 * want to show the progress bar */
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+		gs_app_set_progress (app, 95);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_UPDATE_APPLIED: {
+		/* ensure the state transition to 'updatable' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+		gs_app_set_progress (app, 100);
+		gs_app_set_state (app, AS_APP_STATE_UPDATABLE);
+
+		break;
+	}
+	default:
+		break;
+	}
+
+	/* ensure that a version (and URL) are set */
+	version = gs_app_get_version (app);
+	if (version == NULL || *version == '\0') {
+		version = eos_updater_get_version (priv->updater_proxy);
+		gs_app_set_version (app, version);
+	}
```

Comment:
Don't you also need to connect to `notify::version` then?

Reply:
Hmm, good point. it shouldn't matter most of the times but if we have the GsApp object for the updater and a new upgrade becomes available, this wouldn't update the version. I will change that.